### PR TITLE
feat(nixos module): add daemon options and client script options

### PR DESCRIPTION
**Pull request by jfroche**
Enable user to customize the enqueue script and the daemon's concurrency and error handling options

### CODE REVIEW

File: module.nix [hunk 1]
```diff
@@ -41,12 +41,48 @@ in
       default = "nixbld";
     };
 
+    retryInterval = mkOption {
+      description = lib.mdDoc ''
+        The number of seconds between attempts to run the hook for a package after an initial failure.
+      '';
+      type = types.nullOr types.int;
+      default = null;
```
Comment (Mic92) [inline module.nix:49]:
```suggestion
      type = types.int;
      default = 1;
```

File: module.nix [hunk 2]
```diff
@@ -41,12 +41,48 @@ in
...
+    };
+
+    retries = mkOption {
+      description = lib.mdDoc ''
+        The maximum number of attempts that will be made to run the hook for a package before giving up and dropping the task altogether.
+      '';
+      type = types.nullOr types.int;
+      default = null;
```
Comment (Mic92) [inline module.nix:57]:
```suggestion
      type = types.int;
      default = 5;
```

File: module.nix [hunk 3]
```diff
@@ -41,12 +41,48 @@ in
...
+    };
+
+    concurrency = mkOption {
+      description = lib.mdDoc ''
+        Sets the maximum number of tasks that can be executed simultaneously.
+        By default it is set to 0 which means there is no limit to the number of tasks that can be run concurrently.
+      '';
+      type = types.nullOr types.int;
+      default = null;
```
Comment (Mic92) [inline module.nix:66]:
```suggestion
      type = types.int;
      default = 0;
```

File: module.nix [hunk 4]
```diff
@@ -90,7 +124,10 @@ in
             "async-nix-post-build-hook.socket"
           ];
           serviceConfig = {
-            ExecStart = "${cfg.package}/bin/queued-build-hook daemon --hook ${hook}";
+            ExecStart = "${cfg.package}/bin/queued-build-hook daemon --hook ${hook} " +
+              optionalString (cfg.retryInterval != null) ("--retry-interval ${toString cfg.retryInterval} ") +
+              optionalString (cfg.retries != null) ("--retry-interval ${toString cfg.retries} ") +
+              optionalString (cfg.concurrency != null) ("--concurrency ${toString cfg.concurrency} ");
```
Comment (Mic92) [inline module.nix:130]:
```suggestion
            ExecStart = "${cfg.package}/bin/queued-build-hook daemon --hook ${hook} " +
              "--retry-interval ${toString cfg.retryInterval} " +
              "--retry-interval ${toString cfg.retries} " +
              "--concurrency ${toString cfg.concurrency} ";
```

File: module.nix (hunk 1 above)
Comment (Mic92) [reply module.nix:49]: It's a bit nicer if one cans see the default values straight from the module.

File: test.nix [hunk 5]
```diff
@@ -10,6 +10,9 @@ pkgs.nixosTest
 
       queued-build-hook = {
         enable = true;
+        concurrency = 8;
+        retries = 3;
+        retryInterval = 5;
```
Comment (Mic92) [inline test.nix:15]:
```suggestion
```

File: module.nix (hunk 1 above)
Comment (jfroche) [reply module.nix]: right, I guess we just need to remember that we need to update the defaults both in the code and in the nixos module

File: module.nix [hunk 6]
```diff
@@ -89,8 +136,27 @@ in
...
+                  echo "Environ $key"
+                  export "$key=$(< "$file")"
+                fi
+              done
+            fi
+            exec ${cfg.package}/bin/queued-build-hook daemon --hook ${hook} --retry-interval ${toString cfg.retryInterval} --retry-interval ${toString cfg.retries} --concurrency ${toString cfg.concurrency} 
+          '';
           serviceConfig = {
-            ExecStart = "${cfg.package}/bin/queued-build-hook daemon --hook ${hook}";
+            DynamicUser = true;
```
Comment (Mic92) [inline module.nix:156]: Not strictly needed but looks nicer in a process list. systemd provides a PAM module to resolve those dynamic users/groups.

```suggestion
            DynamicUser = true;
            User = "queued-build-hook";
            Group = "queued-build-hook";
```

**Comment by Mic92** [comment]
bors merge

**Pull request merged**